qiskit_experiments/calibration/calibrations.py

```diff
@@ -0,0 +1,438 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Class to store the results of a calibration experiments."""
+
+import copy
+import dataclasses
+from collections import namedtuple
+from datetime import datetime
+from typing import Any, Dict, Set, Tuple, Union, List, Optional
+
+from qiskit.providers.backend import BackendV1 as Backend
+from qiskit.circuit import Gate
+from qiskit import QuantumCircuit
+from qiskit.pulse import Schedule, DriveChannel, ControlChannel, MeasureChannel
+from qiskit.pulse.channels import PulseChannel
+from qiskit.circuit import Parameter
+from qiskit_experiments.calibration.exceptions import CalibrationError
+from qiskit_experiments.calibration.parameter_value import ParameterValue
+
+ParameterKey = namedtuple("ParameterKey", ["schedule", "parameter"])
+
+
+class Calibrations:
+    """
+    A class to manage schedules with calibrated parameter values.
+    Schedules are stored in a dict and are intended to be fully parameterized,
+    including the index of the channels. The parameter values are stored in a
+    dict where parameters are keys.
+    """
+
+    def __init__(self, backend: Backend):
+        """
+        Initialize the instructions from a given backend.
+
+        Args:
+            backend: The backend from which to get the configuration.
+        """
+
+        self._n_qubits = backend.configuration().num_qubits
+        self._n_uchannels = backend.configuration().n_uchannels
+        self._properties = backend.properties()
+        self._config = backend.configuration()
+        self._params = {}
+
+        # Required because copying a template schedule creates new paramters with new IDs.
+        self._parameter_map = {}
+        self._schedules = {}
+
+    def add_schedules(self, schedules: Union[Schedule, List[Schedule]]):
+        """
+        Add a schedule and register the parameters.
+
+        Args:
+            schedules: The schedule to add.
+
+        Raises:
+            CalibrationError: If the parameterized channel index is not formatted
+                following index1.index2...
+        """
+        if isinstance(schedules, Schedule):
+            schedules = [schedules]
+
+        for schedule in schedules:
+
+            # check that channels, if parameterized, have the proper name format.
+            # pylint: disable = raise-missing-from
+            for ch in schedule.channels:
+                if isinstance(ch.index, Parameter):
```

Comment:
I had a look at `ParameterVector` and I don't think it will do the job. What I would really like is a neat way of linking control channels and drive channels. For example when we do
```python
d0 = Parameter('0')
d1 = Parameter('1')
u10 = Parameter('1.0')

dch0 = DriveChannel(d0)
dch1 = DriveChannel(d1)
uch = ControlChannel(u10)
```
it would appear that we have three degrees of freedom (`d0`, `d1`, `u10`). This is in fact not true as the value of the index of channels `dch0` and `dch1` uniquely determines the index of the control channel. The current implementation solves this issue by imposing a constraint on the name of parameters in control channels. I don't find this solution particularly elegant. Do you have a better suggestion?